File: examples/python/03_custom_backends.py
"""UPP Custom Backends — Implementing Protocol-Based Pluggability.

Demonstrates how to create custom implementations of UPP backend
protocols using Python's structural subtyping (duck typing).

Backend Protocols:
1. IngestBackend — Persists and manages events
2. RetrieverBackend — Intelligent retrieval of relevant events
3. OntologyBackend — Label definitions and server metadata

Any class that implements the required methods satisfies the protocol —
no inheritance needed. This is how UPP enables pluggable backends.

Expected output:
    === Custom Store: LoggingIngest ===
    [LOG] ingest: storing 2 event(s) for user-carol
    Stored 2 events
    [LOG] get_events: retrieving events for user-carol
    Retrieved 2 events
    ...
    === Custom Retriever: KeywordRetriever ===
    Query: "food preferences"
    [RETRIEVER] Searching 2 events for keywords: food, preferences
    Found 1 relevant event(s)
    ...
    === Protocol Compliance Check ===
    LoggingIngest satisfies IngestBackend: True
    KeywordRetriever satisfies RetrieverBackend: True
    ...
"""

import asyncio
import uuid
from datetime import datetime, timezone

from upp import (
    Event,
    EventStatus,
    IngestBackend,
    LabelDefinition,
    OntologyUserV1,
    RetrieverBackend,
    SourceType,
    StoredEvent,
    UPPClient,
)


# =============================================================================
# Custom Backend 1: A logging ingest backend
# =============================================================================


class LoggingIngest:
    """An ingest backend that logs every operation.

    Demonstrates implementing IngestBackend from scratch with logging.
    """

    def __init__(self) -> None:
        self._events: dict[str, list[StoredEvent]] = {}

    async def ingest(
        self, entity_key: str, text: str
    ) -> list[StoredEvent]:
        print(f"  [LOG] ingest: extracting events from text for {entity_key}")
        extracted = self._extract(text)
        print(f"  [LOG] ingest: extracted {len(extracted)} event(s)")
        return await self._persist(entity_key, extracted)

    async def _persist(
        self, entity_key: str, events: list[Event]
    ) -> list[StoredEvent]:
        stored = []
        for e in events:
            se = StoredEvent(
                **e.model_dump(),
                id=str(uuid.uuid4()),
                entity_key=entity_key,
                status=EventStatus.VALID,
                created_at=datetime.now(timezone.utc),
            )
            stored.append(se)
        self._events.setdefault(entity_key, []).extend(stored)
        return stored

    @staticmethod
    def _extract(text: str) -> list[Event]:
        """Naive keyword extraction for demo. In production, use an LLM."""
        events: list[Event] = []
        lower = text.lower()
        if any(w in lower for w in ("food", "spicy", "prefer", "dark mode", "mode")):
            events.append(Event(
                value=text.strip(),
                labels=["what_preferences"],
                confidence=0.92,
                source_type=SourceType.AGENT_OBSERVED if "dark mode" in lower else SourceType.USER_STATED,
            ))
        if any(w in lower for w in ("work", "scientist", "engineer")):
            events.append(Event(
                value=text.strip(),
                labels=["who_role"],
                confidence=0.88,
                source_type=SourceType.USER_STATED,
            ))
        return events

    async def delete_events(
        self, entity_key: str, event_ids: list[str] | None = None
    ) -> int:
        scope = f"event_ids={event_ids}" if event_ids else "all events"
        print(f"  [LOG] delete_events: deleting {scope} for {entity_key}")
        if entity_key not in self._events:
            return 0
        if event_ids is None:
            return len(self._events.pop(entity_key))
        before = len(self._events[entity_key])
        self._events[entity_key] = [
            e for e in self._events[entity_key] if e.id not in set(event_ids)
        ]
        return before - len(self._events[entity_key])

    async def import_events(
        self, entity_key: str, events: list[Event]
    ) -> list[StoredEvent]:
        print(f"  [LOG] import_events: importing {len(events)} event(s) for {entity_key}")
        return await self._persist(entity_key, events)


# =============================================================================
# Custom Backend 2: A keyword-based retriever
# =============================================================================


class KeywordRetriever:
    """A retriever that matches events by keyword overlap.

    Scores events based on how many query keywords appear in the
    event value, then returns matches sorted by score.
    """

    def __init__(self, ingest: LoggingIngest) -> None:
        self._ingest = ingest

    async def retrieve(
        self,
        entity_key: str,
        query: str,
    ) -> list[StoredEvent]:
        stored = self._ingest._events.get(entity_key, [])
        keywords = {w.lower().strip("?.,!") for w in query.split() if len(w) > 2}
        print(f"  [RETRIEVER] Searching {len(stored)} events for keywords: {', '.join(sorted(keywords))}")

        scored: list[tuple[float, StoredEvent]] = []
        for e in stored:
            if e.status != EventStatus.VALID:
                continue
            value_words = {w.lower() for w in e.value.split()}
            overlap = len(keywords & value_words)
            if overlap > 0:
                score = overlap / len(keywords)
                scored.append((score, e))

        scored.sort(key=lambda x: x[0], reverse=True)
        return [event for _, event in scored]

    async def get_events(self, entity_key: str) -> list[StoredEvent]:
        print(f"  [LOG] get_events: retrieving events for {entity_key}")
        return list(self._ingest._events.get(entity_key, []))

    async def export_events(self, entity_key: str) -> list[StoredEvent]:
        print(f"  [LOG] export_events: exporting events for {entity_key}")
        return list(self._ingest._events.get(entity_key, []))


async def main() -> None:
    # =========================================================================
    # 1. Use the custom LoggingIngest
    # =========================================================================
    print("=== Custom Store: LoggingIngest ===")

    ingest = LoggingIngest()

    stored = await ingest.ingest("user-carol", "Loves spicy Thai food")
    stored += await ingest.ingest("user-carol", "Works as a data scientist")
    print(f"Stored {len(stored)} events")

    retriever = KeywordRetriever(ingest)
    retrieved = await retriever.get_events("user-carol")
    print(f"Retrieved {len(retrieved)} events")
    print()

    # =========================================================================
    # 2. Use the custom KeywordRetriever
    # =========================================================================
    print("=== Custom Retriever: KeywordRetriever ===")

    query = "food preferences"
    print(f'Query: "{query}"')

    results = await retriever.retrieve("user-carol", query)
    print(f"Found {len(results)} relevant event(s)")
    for r in results:
        print(f"  - {r.value} {r.labels}")
    print()

    # =========================================================================
    # 3. Protocol compliance check (duck typing)
    # =========================================================================
    print("=== Protocol Compliance Check ===")

    print(f"LoggingIngest satisfies IngestBackend: {isinstance(ingest, IngestBackend)}")
    print(f"KeywordRetriever satisfies RetrieverBackend: {isinstance(retriever, RetrieverBackend)}")
    print()

    # =========================================================================
    # 4. Wire it together with UPPClient
    # =========================================================================
    print("=== Full Client with Custom Backends ===")

    ontology = OntologyUserV1()
    client = UPPClient(ingest=ingest, retriever=retriever, ontology=ontology)

    # Ingest via client
    await client.ingest("user-carol", "Prefers dark mode in all apps")

    # Retrieve via client
    results = await client.retrieve("user-carol", "food and dark mode preferences")
    print(f"Client retrieved {len(results)} events")
    for r in results:
        print(f"  - {r.value}")

    # Server info via client
    info = client.info()
    print(f"\nServer info: protocol v{info['protocol_version']}, "
          f"ontology {info['ontology']}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/03-custom-backends.ts
/**
 * UPP Custom Backends — Implementing Protocol-Based Pluggability.
 *
 * Demonstrates how to create custom implementations of UPP backend
 * interfaces using TypeScript's structural typing.
 *
 * Backend Interfaces:
 * 1. IngestBackend — Persists and retrieves events
 * 2. RetrieverBackend — Intelligent retrieval of relevant events
 * 3. OntologyBackend — Label definitions and server metadata
 *
 * Any class that implements the required methods satisfies the interface —
 * this is how UPP enables pluggable backends.
 *
 * Expected output:
 *     === Custom Store: LoggingStore ===
 *     [LOG] ingestEvents: storing 2 event(s) for user-carol
 *     Stored 2 events
 *     [LOG] getEvents: retrieving events for user-carol
 *     Retrieved 2 events
 *     ...
 *     === Custom Retriever: KeywordRetriever ===
 *     Query: "food preferences"
 *     [RETRIEVER] Searching 2 events for keywords: food, preferences
 *     Found 1 relevant event(s)
 *     ...
 *     === Protocol Compliance Check ===
 *     LoggingStore satisfies IngestBackend: true
 *     KeywordRetriever satisfies RetrieverBackend: true
 *     ...
 */

import { randomUUID } from "node:crypto";

import {
  createEvent,
  createStoredEvent,
  DefaultOntology,
  UppServer,
  UppClient,
  type Event,
  type StoredEvent,
  type IngestBackend,
  type RetrieverBackend,
  type JsonRpcRequest,
  type JsonRpcResponse,
} from "../../implementations/typescript/src/index.js";

// =============================================================================
// Simple InMemoryStore (base for the logging wrapper)
// =============================================================================

class InMemoryStore implements IngestBackend {
  private events: Map<string, StoredEvent[]> = new Map();

  async ingestEvents(entityKey: string, text: string): Promise<StoredEvent[]> {
    const extracted = this.extract(text);
    return this.persist(entityKey, extracted);
  }

  private extract(text: string): Event[] {
    // Naive keyword extraction for demo. In production, use an LLM.
    const events: Event[] = [];
    const lower = text.toLowerCase();
    if (["food", "spicy", "prefer", "dark mode", "mode"].some((w) => lower.includes(w))) {
      events.push(
        createEvent({
          value: text,
          labels: ["what_preferences"],
          confidence: 0.92,
          source_type: lower.includes("dark mode") ? "agent_observed" : "user_stated",
        }),
      );
    }
    if (["work", "scientist", "engineer"].some((w) => lower.includes(w))) {
      events.push(
        createEvent({
          value: text,
          labels: ["who_role"],
          confidence: 0.88,
          source_type: "user_stated",
        }),
      );
    }
    return events;
  }

  persist(entityKey: string, events: Event[]): StoredEvent[] {
    const existing = this.events.get(entityKey) ?? [];
    const stored: StoredEvent[] = [];
    for (const event of events) {
      const se = createStoredEvent({
        ...event,
        id: randomUUID(),
        entity_key: entityKey,
        status: "valid",
        created_at: new Date().toISOString(),
      });
      stored.push(se);
      existing.push(se);
    }
    this.events.set(entityKey, existing);
    return stored;
  }

  async getEvents(entityKey: string, _ontology?: string): Promise<StoredEvent[]> {
    return [...(this.events.get(entityKey) ?? [])];
  }

  async deleteEvents(entityKey: string, eventIds?: string[]): Promise<number> {
    const existing = this.events.get(entityKey) ?? [];
    if (!eventIds) {
      this.events.set(entityKey, []);
      return existing.length;
    }
    const toDelete = new Set(eventIds);
    const remaining = existing.filter((e) => !toDelete.has(e.id));
    this.events.set(entityKey, remaining);
    return existing.length - remaining.length;
  }

  async exportEvents(entityKey: string, _ontology?: string): Promise<StoredEvent[]> {
    const all = await this.getEvents(entityKey);
    return all.filter((e) => e.status === "valid");
  }

  async importEvents(entityKey: string, events: Event[], _ontology?: string): Promise<StoredEvent[]> {
    return this.persist(entityKey, events);
  }
}

// =============================================================================
// Custom Backend 1: A logging wrapper around InMemoryStore
// =============================================================================

class LoggingStore implements IngestBackend {
  private inner = new InMemoryStore();

  async ingestEvents(entityKey: string, text: string): Promise<StoredEvent[]> {
    console.log(`  [LOG] ingestEvents: extracting events from text for ${entityKey}`);
    return this.inner.ingestEvents(entityKey, text);
  }

  async getEvents(entityKey: string, ontology?: string): Promise<StoredEvent[]> {
    console.log(`  [LOG] getEvents: retrieving events for ${entityKey}`);
    return this.inner.getEvents(entityKey, ontology);
  }

  async deleteEvents(entityKey: string, eventIds?: string[]): Promise<number> {
    const scope = eventIds ? `event_ids=[${eventIds.join(", ")}]` : "all events";
    console.log(`  [LOG] deleteEvents: deleting ${scope} for ${entityKey}`);
    return this.inner.deleteEvents(entityKey, eventIds);
  }

  async exportEvents(entityKey: string, ontology?: string): Promise<StoredEvent[]> {
    console.log(`  [LOG] exportEvents: exporting events for ${entityKey}`);
    return this.inner.exportEvents(entityKey, ontology);
  }

  async importEvents(entityKey: string, events: Event[], ontology?: string): Promise<StoredEvent[]> {
    console.log(`  [LOG] importEvents: importing ${events.length} event(s) for ${entityKey}`);
    return this.inner.importEvents(entityKey, events, ontology);
  }
}

// =============================================================================
// Custom Backend 2: A keyword-based retriever
// =============================================================================

class KeywordRetriever implements RetrieverBackend {
  private store: LoggingStore;

  constructor(store: LoggingStore) {
    this.store = store;
  }

  async retrieve(entityKey: string, query: string, _ontology?: string): Promise<Event[]> {
    const stored = await this.store.getEvents(entityKey);
    const keywords = new Set(
      query
        .toLowerCase()
        .split(/\s+/)
        .map((w) => w.replace(/[?.,!]/g, ""))
        .filter((w) => w.length > 2),
    );
    console.log(
      `  [RETRIEVER] Searching ${stored.length} events for keywords: ${[...keywords].sort().join(", ")}`,
    );

    const scored: Array<{ score: number; event: Event }> = [];
    for (const e of stored) {
      if (e.status !== "valid") continue;
      const valueWords = new Set(e.value.toLowerCase().split(/\s+/));
      const overlap = [...keywords].filter((w) => valueWords.has(w)).length;
      if (overlap > 0) {
        scored.push({
          score: overlap / keywords.size,
          event: createEvent({
            value: e.value,
            labels: e.labels,
            confidence: e.confidence,
            source_type: e.source_type,
          }),
        });
      }
    }

    scored.sort((a, b) => b.score - a.score);
    return scored.map((s) => s.event);
  }
}

// =============================================================================
// Main
// =============================================================================

async function main(): Promise<void> {
  // =========================================================================
  // 1. Use the custom LoggingStore
  // =========================================================================
  console.log("=== Custom Store: LoggingStore ===");

  const store = new LoggingStore();

  let stored = await store.ingestEvents("user-carol", "Loves spicy Thai food");
  stored = stored.concat(await store.ingestEvents("user-carol", "Works as a data scientist"));
  console.log(`Stored ${stored.length} events`);

  const retrieved = await store.getEvents("user-carol");
  console.log(`Retrieved ${retrieved.length} events`);
  console.log();

  // =========================================================================
  // 2. Use the custom KeywordRetriever
  // =========================================================================
  console.log("=== Custom Retriever: KeywordRetriever ===");

  const retriever = new KeywordRetriever(store);
  const query = "food preferences";
  console.log(`Query: "${query}"`);

  const results = await retriever.retrieve("user-carol", query);
  console.log(`Found ${results.length} relevant event(s)`);
  for (const r of results) {
    console.log(`  - ${r.value} [${r.labels.join(", ")}]`);
  }
  console.log();

  // =========================================================================
  // 3. Protocol compliance check (structural typing)
  // =========================================================================
  console.log("=== Protocol Compliance Check ===");

  // In TypeScript, structural typing means any object with the right shape
  // satisfies the interface. These checks verify the runtime structure.
  const storeOk: IngestBackend = store; // Compiles → satisfies IngestBackend
  const retrieverOk: RetrieverBackend = retriever; // Compiles → satisfies RetrieverBackend
  console.log(`LoggingStore satisfies IngestBackend: ${storeOk !== undefined}`);
  console.log(`KeywordRetriever satisfies RetrieverBackend: ${retrieverOk !== undefined}`);
  console.log();

  // =========================================================================
  // 4. Wire it together with UppServer + UppClient
  // =========================================================================
  console.log("=== Full Client with Custom Backends ===");

  const ontology = new DefaultOntology();
  const server = new UppServer({
    store,
    retriever,
    ontology,
  });

  // Create an in-process transport (server.handleRequest as transport)
  const transport = async (request: JsonRpcRequest): Promise<JsonRpcResponse> => {
    return server.handleRequest(request);
  };

  const client = new UppClient(transport);

  // Ingest via the backend directly
  await store.ingestEvents("user-carol", "Prefers dark mode in all apps");

  // Retrieve via client
  const clientResults = await client.retrieve("user-carol", "food and dark mode preferences");
  console.log(`Client retrieved ${clientResults.length} events`);
  for (const r of clientResults) {
    console.log(`  - ${r.value}`);
  }

  // Server info via client
  const info = await client.info();
  console.log(
    `\nServer info: protocol v${info.protocol_version}, ${info.operations.length} operations`,
  );
}

main().catch(console.error);
